WalletWasabi/Extensions/NBitcoinExtensions.cs

```diff
@@ -134,6 +135,36 @@ public static int GetAnonymitySet(this Transaction me, int outputIndex)
 
 		public static int GetAnonymitySet(this Transaction me, uint outputIndex) => GetAnonymitySet(me, (int)outputIndex);
 
+		public static int GetAnonymitySet(this Transaction me, int outputIndex, IEnumerable<SmartCoin> spentOwnCoins)
+		{
+			// Get the anonymity set of i-th output in the transaction.
+			var anonset = me.GetAnonymitySet(outputIndex);
+			var numberOfOwnInputs = spentOwnCoins.Count();
+			// If we provided inputs to the transaction.
+			if (numberOfOwnInputs > 0)
+			{
+				// Take the input that we provided with the smallest anonset.
+				// Our smallest anonset input is the relevant here, because this way the common input ownership heuristic is considered.
+				var smallestInputAnon = spentOwnCoins.Min(x => x.AnonymitySet);
+
+				// Punish consilidation exponenetially.
+				// If there is only a single input then the exponent should be zero to divide by 1 thus retain the input coin anonset.
+				var consolidatePenalty = Math.Pow(2, numberOfOwnInputs - 1);
+				var privacyBonus = smallestInputAnon / consolidatePenalty;
+
+				// If the privacy bonus is <=1 then we are not adding any privacy to the coin.
+				var normalizedBonus = privacyBonus - 1;
```

Comment:
hmm, what does the -1 account for here?

Reply:
Imagine I am spending a coin of mine to myself that has 1 anonset. In that case, without - 1, the algorithm would throw out anonset 2 for the output instead of 1. That's what's accounted here.

Reply:
ah, makes sense. can resolve